src/platform/plugins/shared/dashboard/public/dashboard_api/esql_variables_manager.ts

```diff
@@ -7,30 +7,64 @@
  * License v3.0 only", or the "Server Side Public License, v 1".
  */
 
-import { BehaviorSubject } from 'rxjs';
+import deepEqual from 'fast-deep-equal';
+
+import type { DefaultEmbeddableApi } from '@kbn/embeddable-plugin/public';
 import type { PublishesESQLVariable } from '@kbn/esql-types';
 import { apiPublishesESQLVariable, type ESQLControlVariable } from '@kbn/esql-types';
 import { combineCompatibleChildrenApis } from '@kbn/presentation-containers';
-import type { DefaultEmbeddableApi } from '@kbn/embeddable-plugin/public';
 import type { PublishingSubject } from '@kbn/presentation-publishing';
+import { BehaviorSubject, combineLatestWith, filter, first, skip } from 'rxjs';
+import type { initializeSettingsManager } from './settings_manager';
 
 export const initializeESQLVariablesManager = (
-  children$: PublishingSubject<{ [key: string]: DefaultEmbeddableApi }>
+  children$: PublishingSubject<{ [key: string]: DefaultEmbeddableApi }>,
+  settingsManager: ReturnType<typeof initializeSettingsManager>
 ) => {
-  const esqlVariables$ = new BehaviorSubject<ESQLControlVariable[]>([]);
+  const publishedEsqlVariables$ = new BehaviorSubject<ESQLControlVariable[]>([]);
+  const unpublishedEsqlVariables$ = new BehaviorSubject<ESQLControlVariable[]>([]);
+
   const childrenESQLVariablesSubscription = combineCompatibleChildrenApis<
     PublishesESQLVariable,
     ESQLControlVariable[]
   >({ children$ }, 'esqlVariable$', apiPublishesESQLVariable, []).subscribe((newESQLVariables) => {
-    esqlVariables$.next(newESQLVariables);
+    unpublishedEsqlVariables$.next(newESQLVariables);
   });
 
+  const publishVariables = () => {
```

Comment:
Excited to have a centralized `composable_fetch_context` implementation as a followup to avoid this duplication. But as discussed offline this is how we'll leave it for the time being.